lxd/api_project.go

```diff
@@ -316,19 +324,36 @@ func projectPatch(d *Daemon, r *http.Request) response.Response {
 		req.Config["features.images"] = project.Config["features.profiles"]
 	}
 
+	_, err = reqRaw.GetString("limits.containers")
+	if err != nil {
+		req.Config["limits.containers"] = project.Config["limits.containers"]
+	}
+
+	_, err = reqRaw.GetString("limits.virtual-machines")
+	if err != nil {
+		req.Config["limits.virtual-machines"] = project.Config["limits.virtual-machines"]
+	}
+
 	return projectChange(d, project, req)
 }
-
 // Common logic between PUT and PATCH.
 func projectChange(d *Daemon, project *api.Project, req api.ProjectPut) response.Response {
 	// Flag indicating if any feature has changed.
 	featuresChanged := req.Config["features.images"] != project.Config["features.images"] || req.Config["features.profiles"] != project.Config["features.profiles"]
+	// Flag indicating if any limit has changed.
+	containerLimitChanged := req.Config["limits.containers"] != project.Config["limits.containers"]
+	vmLimitChanged := req.Config["limits.virtual-machines"] != project.Config["limits.virtual-machines"]
+	limitsChanged := containerLimitChanged || vmLimitChanged
```

Comment:
Feels like all of this could be made a whole lot cleaned by having a function which compares `req.Config` and `project.Config` and returns a list of keys that have changed and a boolean to indicate whether features have been changed (as needed in a few of the checks below).